modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj

```diff
@@ -146,6 +154,65 @@
                table-schema->metabase-field-info
                set)})
 
+(def ^:private ^DateTimeFormatter bq-ts-str-formatter
+  "For some crazy reason, the TIMESTAMP String values returned from [[FieldValue]] look like `1577836800.0`, so create
+  a parser for that.  This is basically '<epoch seconds>.<millis>'."
+  (.toFormatter (doto (DateTimeFormatterBuilder.)
+                  (.appendValue ChronoField/INSTANT_SECONDS 1 19 SignStyle/NEVER)
+                  (.appendLiteral \.)
+                  (.appendValue ChronoField/MILLI_OF_SECOND))))
+
+(defn- bq-ts-str->instant [ts-str]
+  (let [parsed (.parse bq-ts-str-formatter ts-str)]
+    (Instant/from parsed)))
+
+(defn- FieldValueList->fingerprint-structure [fields truncation-size ^FieldValueList values]
+  (map-indexed (fn [idx ^FieldValue fv]
+                 (let [{:keys [:database_type]} (nth fields idx)
+                       v (.getValue fv)]
+                      (if (string? v)
+                        (cond (= "TIMESTAMP" database_type)
+                              (bq-ts-str->instant v)
+
+                              (some? truncation-size)
+                              (let [^String str-val v]
+                                (.substring ^String str-val 0 (min truncation-size (.length str-val))))
+
+                              true
+                              v)
+                        v)))
+               ;; the list operation will return *all* field values for each row, in the database field order
+               ;; so, map over the requested fields and fetch only those values at each respective :database_position
+               (map #(-> values
+                         (.get (:database_position %))
+                         (.getValue))
+                    fields)))
+
+;; Override the table-rows-sample multimethod for :bigquery-cloud-sdk, to use a different mechanism for sampling
+;; table rows (the list operations, rather than a regular query), purely as a cost control measure.  For certain types
+;; of tables, BigQuery evidently does not honor the limit parameter of a query, and instead performs a full table scan
+;; and then truncates results at the end.  See #19470 for full details.
+(defmethod @#'sync.f/table-rows-sample :bigquery-cloud-sdk [driver
+                                                            {table-name :name, dataset-id :schema :as table}
+                                                            fields
+                                                            rff
+                                                            opts]
+  (let [database (table/database table)
+        bq-tbl   (get-table database dataset-id table-name)
+        tbl-type (-> (.getDefinition bq-tbl)
+                     (.getType))]
+    (if (contains? #{TableDefinition$Type/VIEW TableDefinition$Type/MATERIALIZED_VIEW} tbl-type)
```

Comment:
This is a good catch. But it makes me worry that there are other types we need to fall back on? Should we instead check that the table definition is instead a type that definitely supports the table sample api? Or is it true all table types except for materialized views are supported?

Reply:
Good questions, which I haven't been able to answer definitively by looking at the API docs.  This particular limitation was discovered empirically (it doesn't appear to be mentioned in the docs).